core/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java

```diff
@@ -60,8 +60,10 @@
     public static final ParseField TYPE_FIELD = new ParseField("type");
     public static final ParseField QUERY_FIELD = new ParseField("query");
 
-    /** The default name for the match query */
-    public static final String NAME = "match";
+    /** The names for the match query. */
+    // TODO replace this with a Types like object
+    public static final Names NAMES = new Names("match", "match_phrase", "matchPhrase", "match_phrase_prefix", "matchPhrasePrefix",
```

Comment:
ok I hadn't looked closely at the match query code before commenting above, this sounds like it should be fixed regardless. But let's keep an eye at backwards compatibility for the query DSL, we don't want to break it without deprecating things first. I would move this match query discussion to a separate issue/PR (that Colin is going to send I think).

Back to the `Names` class discussion, I think there are two main concerns:
1) we should use `ParseField` for parsing. We already do in the match query partially, see `MatchQueryBuilder#MATCH_PHRASE_FIELD` etc., so we end up duplicating names like `phrase_prefix` etc. I don't think we should be concerned about moving the camel case variants into ParseField that handles them automatically, I don't consider that too much magic. We should actually do that so that we can centralize where camel case is handled and get ready for removing support for it without having to touch every single query. This makes me lean more towards Colin's idea that each query should ideally have one single name (the preferred one) and all the others are just camel case variants that `ParseField` handles internally. The exceptions are:
- match: it needs fixing, we'll discuss that separately
- more_like_this: it can also be referred to as `mlt`. I think the synonym should/can be deprecated
- terms: it can also be referred to as `in`. This last synonym should probably be deprecated
- geo_bbox: can be be referred to as `geo_bounding_box`. I think one of the variants should be deprecated.

This seems to me like a perfect usecase for `ParseField` then once we addressed these few issues above.

2) we want to keep serialization distinct from parsing, although we use the same query id in both. That could be achieved by using ParseField within Names maybe? I tend to agree that using ParseField and call getPreferredName against it for serialization could be a misuse.

Let me know your thoughts.